### ISSUE TITLE
SIGABRT when calling oneapi::tbb::finalize()

### ISSUE THREAD

**Issue by rsandvik** [oneapi-src/oneTBB#1040]
We have the following issue on linux Ubuntu 20.04.5 (64x86):

We load a DLL (librun.so) from a python script (test.py) using cpython. The DLL librun uses TBB and runs some simple parallel constructs in a task arena. When librun is unloaded we call `oneapi::tbb::finalize()` from the exit handler of librun. This results in `SIGABRT`. GDB back-trace below.

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7dc1859 in __GI_abort () at abort.c:79
#2  0x00007ffff71d23df in __gnu_cxx::__verbose_terminate_handler () at /opt/conda/conda-bld/compilers_linux-64_1534514838838/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007ffff71d0b16 in __cxxabiv1::__terminate (handler=<optimized out>) at /opt/conda/conda-bld/compilers_linux-64_1534514838838/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007ffff71d0b4c in std::terminate () at /opt/conda/conda-bld/compilers_linux-64_1534514838838/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00007ffff71d1516 in __cxxabiv1::__cxa_pure_virtual () at /opt/conda/conda-bld/compilers_linux-64_1534514838838/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/libsupc++/pure.cc:50
#6  0x00007ffff754efe8 in tbb::detail::r1::global_control_impl::destroy (gc=...) at /tmp/tbb/oneTBB/src/tbb/global_control.cpp:213
#7  tbb::detail::r1::destroy (gc=...) at /tmp/tbb/oneTBB/src/tbb/global_control.cpp:250
#8  0x00007ffff7558656 in tbb::detail::d1::global_control::~global_control (this=<optimized out>, this=<optimized out>) at /tmp/tbb/oneTBB/src/tbb/../../include/oneapi/tbb/global_control.h:87
#9  tbb::detail::r1::release_impl (handle=...) at /tmp/tbb/oneTBB/src/tbb/governor.cpp:267
#10 tbb::detail::r1::finalize (mode=<optimized out>, handle=...) at /tmp/tbb/oneTBB/src/tbb/governor.cpp:305
#11 tbb::detail::r1::finalize (handle=..., mode=<optimized out>) at /tmp/tbb/oneTBB/src/tbb/governor.cpp:298
#12 0x00007ffff758c159 in tbb::detail::d1::finalize(tbb::detail::d1::task_scheduler_handle&)::{lambda()#1}::operator()() const (__closure=0x7fffffffe410) at include/tbb/../oneapi/tbb/global_control.h:168
#13 0x00007ffff758ccbc in tbb::detail::d0::try_call_proxy<tbb::detail::d1::finalize(tbb::detail::d1::task_scheduler_handle&)::{lambda()#1}>::on_completion<tbb::detail::d1::finalize(tbb::detail::d1::task_scheduler_handle&)::{lambda()#2}>(tbb::detail::d1::finalize(tbb::detail::d1::task_scheduler_handle&)::{lambda()#2}) (this=0x7fffffffe410, on_completion_body=...) at include/tbb/../oneapi/tbb/detail/_template_helpers.h:230
#14 0x00007ffff758c1aa in tbb::detail::d1::finalize (handle=...) at include/tbb/../oneapi/tbb/global_control.h:172
#15 0x00007ffff758a9d9 in mosek_sofinalize () at run.cc:42
#16 0x00007ffff7fe0f6b in _dl_fini () at dl-fini.c:138
#17 0x00007ffff7de58a7 in __run_exit_handlers (status=0, listp=0x7ffff7f8b718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#18 0x00007ffff7de5a60 in __GI_exit (status=<optimized out>) at exit.c:139
#19 0x00007ffff7dc308a in __libc_start_main (main=0x5555556496c0 <main>, argc=2, argv=0x7fffffffe6a8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe698) at ../csu/libc-start.c:342
#20 0x000055555572f062 in _start () at ../sysdeps/x86_64/elf/start.S:103
```
Code to reproduce attached below. To compile and run use `./compile.sh && ./run.sh`.

[finalize_crash.tar.gz](https://github.com/oneapi-src/oneTBB/files/10935417/finalize_crash.tar.gz)

To me it seems like the virtual function `control_storage::default_value()` is being called on an object for which the vtable has been freed. 

run.cc
```
#include "tbb/global_control.h"
#include "tbb/parallel_invoke.h"
#include "tbb/parallel_for.h"
#include "tbb/task_group.h"
#include "tbb/task_arena.h"

#include <iostream>
#include <mutex>

auto arena = oneapi::tbb::task_arena(64);

extern "C" void run()
{
    arena.initialize();
    std::cout << "Implementation of \"" << __func__ << "()\"" << std::endl;

    arena.execute(
        []
        {
            std::mutex m;

            tbb::parallel_for(tbb::blocked_range<int>(0, 128),
                              [&](tbb::blocked_range<int> r)
                              {
                                  for (int i = r.begin(); i < r.end(); ++i)
                                  {
                                      const std::lock_guard<std::mutex> lock(m);
                                      std::cout << "Iteration #" << i << std::endl;
                                  }
                              });
        });
}

__attribute__((destructor)) static void mosek_sofinalize(void)
{
    std::cout << "Calling arena.terminate();" << std::endl;
    arena.terminate();
    oneapi::tbb::task_scheduler_handle handle = oneapi::tbb::task_scheduler_handle{oneapi::tbb::attach{}};
    try
    {
        std::cout << "Calling oneapi::tbb::finalize(handle);" << std::endl;
        oneapi::tbb::finalize(handle);
        // oneTBB worker threads are terminated at this point.
    }
    catch (const oneapi::tbb::unsafe_wait &)
    {
        std::cerr << "Failed to terminate the worker threads." << std::endl;
    }

    return;
}
```

test.py
```
from ctypes import  *

run = cdll.LoadLibrary("librun.so")
run.run()
print("Exiting python script")
```

**Comment by JhaShweta1** [comment]
hi @rsandvik, there are a couple of solutions suggested here https://github.com/oneapi-src/oneTBB/issues/977. Could you try them and let us know if you still face the issue.

**Comment by rsandvik** [comment]
Hi, 

I have read #977 and this issue is a result of trying to apply the suggestions in it. Not calling `finalize()` in the code above results in a crash which was what was reported in #977. As suggested we try to call `finalize()` before the DLL is unload.  This also results in a crash if called when the DLL is unloaded as in the provided example. If I instead call `finalize()` before the function `run()` exits the code does not crash. So my question is, should it be possible to call `finalize()` when the DLL is unloaded without it resulting in an error?

**Comment by pavelkumbrasev** [comment]
Hi @rsandvik ,
`finalize` uses library entry point on creation (not default one) so if library is already unloaded it always would lead to the crush.  
To be honest, I still don't understand your usage example. Why `librun` is unloaded before oneTBB finalization? Is there a way to control it?
Also, `finalize` before `run` seems to be NOP because on `run` TBB will initialize library again.

**Comment by erling-d-andersen** [comment]
You have a complete smallish example that illustrates a crash that mimics a real use case.  Is what we are doing wrong? If yes, then what do we do wrong? 

We have access to Intel compiler support. Maybe we should post it there.

We first posted it

https://community.intel.com/t5/Intel-oneAPI-Threading-Building/Segmentation-fault-using-dlopen-dlclose-on-Linux/m-p/1457804/emcs_t/S2h8ZW1haWx8dG9waWNfc3Vic2NyaXB0aW9ufExFRFNXSkFLN1lYR0NWfDE0NTc4MDR8U1VCU0NSSVBUSU9OU3xoSw#M15332

where a small example was required. We have now made it available.

**Comment by pavelkumbrasev** [comment]
@erling-d-andersen a I understand from forum thread your problem is covered in https://github.com/oneapi-src/oneTBB/issues/977
So, if you can call `finalize` before library unload everything should be fine and it will not have any side effect on other libraries that use  oneTBB (I have tried to explain how this is work in https://github.com/oneapi-src/oneTBB/issues/977)
This is issue related to oneTBB so if you still have a problem after applying this solution please create new issue in our repo.

**Comment by aszekMosek** [comment]
@pavelkumbrasev We have a library that is linked with TBB. Our library, in turn, can be loaded by others, in particular by third parties using ``dlopen`` and by Python, Matlab, java, you name it.

If we don't call finalize() then we run into the problem of #977 when our library is used via ``dlopen``.

If we call finalize() in the unload code of our library then a Python, Java, etc. program using our library is terminated, as demonstrated by the simplified example here. We have no control over the order in which various libraries are terminated/unloaded when the Python/java etc. program is stopped. For this reason we cannot really call finalize() prior to unloading, because we cannot for example force the Python interpreter or the Java runtime to do it.

I hope that explains the condurum.

**Comment by pavelkumbrasev** [comment]
@aszekMosek Thank you for the explanation it is clear now!
Seems to be a really tricky problem!
Whilst, I don't know a clear solution but I think I can suggest a workaround for this problem.
In your case you rely on work of the OS linker and oneTBB could be unloaded before your library will be unloaded.
What if you will link with oneTBB dynamically using `dlopen/dlclose` so you can control oneTBB lifetime and `finalize` will become safe then.
May be there are some issues that I don't take into consideration?

**Comment by erling-d-andersen** [comment]
Thanks for the feedback. This is really a major headache for us. Until now only one customer has complained until now.
However, we think it is luck only and as a commercial library vendor we rely on oneTBB actually works in all circumstances.
To the best of my knowledge we are not doing anything wrong according to the oneTBB documentation.

Would Iinking oneTBB static solve it?

**Comment by pavelkumbrasev** [comment]
Actually, I think we missed something because it is really strange behavior when library1 -> `oneTBB` unloaded before library2 -> `your library` that have dependency on library1 I believe OS or runtime loader should resolve such dependencies and library2 should always be unloaded after library1 because it has dependency.
So generally I think there is some problem that we don't understand.
In my mind it can be solved by workaround. There are more options:

- Link with oneTBB as usual but on entry point to your library add extra reference by dlopen (it should support such behavior) and on exit point call dlclose on oneTBB handle.
- Call dlopen before 'finalize' and dlclose after. (There is should be an option that return handler only if library already loaded in application virtual space so if nullptr returned there is no need to call `finalize`)
- Check for some oneTBB symbol if it is still in process virtual space and call finalize only if oneTBB still there.

These option are tricky but I think this is only because something goes wrong with loader dependencies.

While, static linking will solve this particular problem it will bring new problems starting with possible symbols conflict ending with 2 thread pools one from statically linked oneTBB and another from dynamically so we usually not recommend to link statically against oneTBB.

**Comment by erling-d-andersen** [comment]
My comments:

* The example we have provided has not been shown to contain bugs nevertheless it crashes.
* To me it seems no fool proof solution has been suggested. 
* Currently, I think the static link is the best solution for us 8not tried yet). Since we carefully control which symbols are exported from our DLL/SO we should not have the symbol conflict issues. Note we link other libraries and customer has to be able to link with incompatible versions of those already.
* Given the status of oneTBB I am bit surprised this issue is not given more attention. (We have a support agreement for Intel C, so I can submit an issue if that helps.)

**Comment by pavelkumbrasev** [comment]
As I noted is not something that oneTBB can control on it side. There is a problem on loader side because dependency library should not be unloaded before dependent library it should not be like this in general not only for oneTBB.
Why are all of the proposed solution unreliable? They are mostly target on forcing the linker to preserve a reference on oneTBB and unload it in safe manner.  
There is still a problem if the library customers links with TBB dynamically there will be two thread pool that might make worse the overall performance.

**Comment by erling-d-andersen** [comment]
I am not saying all the proposed solutions are unreliable. However, can you say with 100% certainty that just one of them is 100% foolproof?

The two thread pool issue is "a feature" and not "a bug" from our and our customers perspective. So that is a vastly different issue than crashes in some circumstances. Hence, it is something we can live with.

**Comment by pavelkumbrasev** [comment]
> I am not saying all the proposed solutions are unreliable. However, can you say with 100% certainty that just one of them is 100% foolproof?

I would say it will require testing on your side but all of them should solve this problem to the best of my knowledge.

**Comment by rsandvik** [comment]
I tried loading / unloading tbb with `dlopen() / dlcolse()` as suggested. See code below. This still crashes with the output:

```
dlopen() TBB

Exiting python script
Calling oneapi::tbb::finalize(handle);
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
```

It seems the crash happens before tbb is unloaded. I don't think the issue is that `oneapi::tbb::finalize()` is called after TBB is unloaded. Rather the issue is that  calling `finalize()` leads to calling a virtual function on a statically allocated object. As far as I know this is not safe to do from the library destructor as the VTABLE for such statically allocated object may have been unloaded even though the library has not.

```c
#include "tbb/global_control.h"
#include "tbb/parallel_invoke.h"
#include "tbb/parallel_for.h"
#include "tbb/task_group.h"
#include "tbb/task_arena.h"
#include <stdio.h>
#include <assert.h>
#include <unistd.h>
#include <dlfcn.h>
#include <ctype.h>

typedef void *soHandle_t;

static soHandle_t loadLib(const char *libName, char **errMsg)
{
    soHandle_t h;

    (void)dlerror();
    h = dlopen(libName, RTLD_NOW | RTLD_GLOBAL);
    if (NULL == h)
    {
        *errMsg = dlerror();
    }
    else
    {
        *errMsg = NULL;
    }

    return h;
} /* loadLib */

static int unLoadLib(soHandle_t hh)
{
    return dlclose(hh);
} /* unLoadLib */

#include <iostream>
#include <mutex>

typedef void *soHandle_t;

static soHandle_t tbblib = NULL;

extern "C" int run()
{
    return 0;
}
__attribute__((constructor)) static void mosek_soconsturct(void)
{
    char *errMsg;
    tbblib = loadLib("libtbb.so", &errMsg);
    std::cout << "dlopen() TBB\n"
              << std::endl;
    if (!tbblib)
        std::cout << "Error Loading tbb\n"
                  << std::endl;
}

__attribute__((destructor)) static void mosek_sofinalize(void)
{
    oneapi::tbb::task_scheduler_handle handle = oneapi::tbb::task_scheduler_handle{oneapi::tbb::attach{}};
    try
    {
        std::cout << "Calling oneapi::tbb::finalize(handle);" << std::endl;
        oneapi::tbb::finalize(handle);
        // oneTBB worker threads are terminated at this point.
    }
    catch (const oneapi::tbb::unsafe_wait &)
    {
        std::cerr << "Failed to terminate the worker threads." << std::endl;
    }
    std::cout << "dlclose() TBB\n"
              << std::endl;
    unLoadLib(tbblib);
    return;
}
```

**Comment by pavelkumbrasev** [comment]
It is unclear for me how is this even possible if you increase the reference on TBB by dlopen in constructor how related to TBB static objects might be freed before dlclose call?

**Comment by rsandvik** [comment]
As far as I can tell, the issue is that the code assumes that destructors of globally allocated objects are called after library destructores. But this ordering is undefined. The attached example demonstrates this. 

dll2.cc
```c
#include <iostream>
#include <cstdlib>

class Base {
public:
    virtual void sayHello() = 0;
    virtual ~Base() {}
};

class Derived : public Base {
public:
    void sayHello() override { std::cout << "Hello from Derived" << std::endl; }
    ~Derived() { std::cout << "Destructor of Derived called" << std::endl; }
};

static Derived globalObj;
Base* globalObjPtr = &globalObj;

void libraryfunction(void)
{
    globalObjPtr->sayHello();
}

void __attribute__((destructor)) myDestructor() {
    std::cout << "Calling dll2 destructor" << std::endl;
}
```

ddl1.cc
```c
#include <iostream>

void libraryfunction(void);

void __attribute__((destructor)) myDestructor() {
    std::cout << "Calling dll1 destructor" << std::endl;
    libraryfunction();
}
```
If dll1 depends on dll2 and you load dll1 you get. 

```
exiting python script
Destructor of Derived called
Calling dll1 destructor
pure virtual method called
terminate called without an active exception
```
So `~Derived()` is called before the library destructor resulting in a virtual function being called an a destructed object, which is undefined behavior. 

As fart as I can tell this example is analog to what is happening in global_controll.cpp.

[ex2.tar.gz](https://github.com/oneapi-src/oneTBB/files/11087012/ex2.tar.gz)

**Comment by pavelkumbrasev** [comment]
Hi @rsandvik, do you have any updates for this issue?

**Comment by erling-d-andersen** [comment]
@rsandvik is on vacation these days.  You can close and we will reopen if needed.

**Comment by ulfworsoe** [comment]
I can add to the thread that this issue also occurs with Matlab with the latest oneTBB (2021.11).

Build a library `mylib` that uses tbb and has a function `fun` that initializes a thread pool.

Basically: use `loadlibrary` to load it, then `calllib` to call the function, then `unloadlibrary` to unload it: Segfauilt.

This also crashes: 
- `loadlibrary("libtbb")`
- `loadlibrary("mylib")`
- `calllib("mylib","fun")`
- `unloadlibrary("mylib")`: now it work, no crashes, but...
- `unloadlibrary("libtbb")`: Segfault.

`mylib`s destructor cannot do `finalize()` since there may be other libraries using the same libtbb.

**Comment by dnmokhov** [comment]
In this case, you can try using several `task_scheduler_handle` instances with a combination of `release()` and `finalize()`:

https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Initializing_and_Terminating_the_Library.html
